ls1intum/Artemis
# `Development`: Reduce client main bundle size and introduce caching

### Checklist
#### General


#### Client

### Motivation and Context



The current main bundle size is 539 KB (gzipped). In general, we should try to reduce the bundle size wherever possible to speed up the application load time.

### Description


This PR works towards reducing the main bundle size and speeding up the build process ([now 532 KB](https://bamboo.ase.in.tum.de/browse/ARTEMIS-WEBAPP4169-11/artifact/shared/webpack-bundle-report/stats.html)).
The dependency `papaparse` is now lazy loaded and was removed from the main bundle (required a workaround because of some weird behaviour when importing it in an abstract superclass).
It also optimizes the imports from `ngx-charts` and `crypto-js` by limiting it to the modules that are actually used.
Ideally, using tree-shaking the Angular Ivy engine should now be able to throw out any modules not used in Artemis.
This [does not seem to work](https://github.com/swimlane/ngx-charts/issues/1765) for `ngx-charts`, yet.

Most importantly, this PR introduces caching for all `*.js, *.css` files (as well as for `/i18n`).
The caching period can be configured using `jhipster.http.cache.timeToLiveInDays`.
Due to [an issue in JHipster](https://github.com/jhipster/generator-jhipster/issues/19102), the server currently does not set the caching period correctly (milliseconds instead of sections, factor 1000).
Because all JS and CSS files are delivered with a hash in the filename, this shouldn't be a big issue and cache busting should work automatically.

Thanks to @akesfeden, @willergeorg and @bassner for their discussion/inputs on this PR.

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Navigate through different components using charts
3. Ensure that all charts are still rendered without issues
4. Ensure all components related to the “grading key” work correctly
5. Ensure that the resource files (`*.js, *.css, /i18n`) are delivered with a `Cache-Control` header
6. Deploy a different PR to the test server to ensure that cache busting works (you should _not_ have to force-reload, usually Shift+F5, to receive the updated resource files)

### Review Progress





#### Code Review
#### Manual Tests
